src/solvers/refinement/string_refinement.cpp

```diff
@@ -1716,9 +1744,186 @@ bool string_refinementt::is_axiom_sat(
+    {
+      // This represents that the expr is a valid leaf, may not be future proof
+      // or 100% enforced, but is correct prescriptively. All non-sum exprs must
+      // be leaves.
+      correct&=(expr.operands().size()==0);
+    }
+  }
+};
+
+/// \related string_constraintt
+static bool is_linear(const exprt &expr)
+{
+  is_linear_visitort v;
+  expr.visit(v);
+  return v.correct;
+}
+
+/// Checks the data invariant for \link string_constraintt
+/// \related string_constraintt
+/// \param [in] expr: the string constraint to check
+/// \return whether the constraint satisfies the invariant
+bool string_refinementt::is_valid_string_constraint(
+  const string_constraintt &expr)
+{
+  // Condition 1: The premise cannot contain any string indices
+  const array_index_mapt premise_indices=gather_indices(expr.premise());
+  if(!premise_indices.empty())
+  {
+    std::string print;
+    for(const auto &pair : premise_indices)
+    {
+      print.append(from_expr(ns, "", pair.first)+": {");
+      for(const auto &i : pair.second)
+        print.append(from_expr(ns, "", i) + ", ");
+      print.append("}");
+    }
+    error() << "Premise has indices: " << from_expr(ns, "", expr.premise())
+            << " => " << from_expr(ns, "", expr.body()) << ", map: {"
+            << print << "}" << eom;
+    return false;
+  }
+
+  const array_index_mapt body_indices=gather_indices(expr.body());
+  // Must validate for each string
+  for(const auto &pair : body_indices)
+  {
+    // Condition 2: All indices of the same string must be the of the same form
+    std::vector<exprt> indices(pair.second);
+    const exprt rep=indices.back();
+    indices.pop_back();
+    for(const auto &i : indices)
+    {
+      const equal_exprt equals(rep, i);
+      const exprt result=simplify_expr(equals, ns);
+      if(result.is_false())
+      {
+        error() << "Indices not equal: " << from_expr(ns, "", expr.premise())
+                << " => " << from_expr(ns, "", expr.body()) << ", str: "
+                << from_expr(ns, "", pair.first) << eom;
+        return false;
+      }
+    }
+
+    // Condition 3: f must be linear
+    if(!is_linear(rep))
+    {
+      error() << "f is not linear: " << from_expr(ns, "", expr.premise())
+              << " => " << from_expr(ns, "", expr.body()) << ", str: "
+              << from_expr(ns, "", pair.first) << eom;
+      return false;
+    }
+
+    // Condition 4: the quantified variable can only occur in indices in the
+    // body
+
+    // We need to check that each instance of the universal variable occurs in
+    // and index expression. For efficiency, we do a depth-first search of the
+    // body. The exprt visitors do a BFS and hide the stack/queue, so we would
+    // need to store a map from child to parent.
+
+    // The int represents what child we will visit next.
+    typedef std::pair<exprt, unsigned> valuet;
+    std::stack<valuet> stack;
+    stack.push(valuet(expr.body(), 0));
```

Comment:
it seems to me that it would make more sense to have the depth as the second element of the pair, and you don't need to store the index of the child if you push all the operands in the stack.